src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -75,138 +74,233 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
+
+        private static void ModifyUnmanagedDllResolver(Action<AssemblyLoadContext> resolverAction)
+        {
+            // For Windows hostfxr is loaded in the process.
+            if (!IsWindows)
             {
-                // 32-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT(x86)", exeName);
+                var loadContext = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+                if (loadContext != null)
+                {
+                    resolverAction(loadContext);
+                }
             }
-            else if (IntPtr.Size == 8)
+        }
+
+        private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
+        {
+            var hostFxrLibName = "libhostfxr";
+            var libExtention = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "dylib" : "so";
+
+            if (!hostFxrLibName.Equals(libraryName))
             {
-                // 64-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT", exeName);
+                return IntPtr.Zero;
             }
 
-            if (dotnetPath is null)
+            var hostFxrRoot = Path.Combine(DotnetPath.Value, "host", "fxr");
+            if (Directory.Exists(hostFxrRoot))
             {
-                // We will generally find the dotnet exe on the path, but on linux, it is often just a 'dotnet' symlink (possibly even to more symlinks) that we have to resolve
-                // to the real dotnet executable.
-                // This will work as often as just invoking dotnet from the command line, but we can be more confident in finding a dotnet executable by following
-                // https://github.com/dotnet/designs/blob/main/accepted/2021/install-location-per-architecture.md
-                // This can be done using the nethost library. We didn't do this previously, so I did not implement this extension.
-                foreach (string dir in Environment.GetEnvironmentVariable("PATH").Split(Path.PathSeparator))
+                // Load hostfxr from the highest version, because it should be backward-compatible
+                var hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
+                            .Max(str => SemanticVersionParser.TryParse(str, out var version) ? version : null);
+
+                if (hostFxrAssemblyDirectory != null && !string.IsNullOrEmpty(hostFxrAssemblyDirectory.OriginalValue))
                 {
-                    string filePath = Path.Combine(dir, exeName);
-                    if (File.Exists(filePath))
+                    var hostfxrAssembly = Directory.GetFiles(hostFxrAssemblyDirectory.OriginalValue)
+                        .FirstOrDefault(filePath => filePath.Equals(Path.Combine(hostFxrLibName, libExtention)));
+
+                    if (hostfxrAssembly != null)
```

Comment:
This looks wrong.
- `Path.Combine("name", "ext")` returns `name/ext`.
- It is wasteful to create a new string on each lambda invocation.
- Now that the comparison is case sensitive, there doesn't seem to be a reason to enumerate - we know what the expected name is.

```suggestion
                    var hostfxrAssembly = Path.Combine(hostFxrAssemblyDirectory.OriginalValue, Path.ChangeExtension(hostFxrLibName, libExtension));
                    if (File.Exists(hostFxrAssembly))
```